File: src/lib/feature-flag-service.ts
import { apiClient } from "./api-client";

export interface FeatureFlag {
  name: string;
  enabled: boolean;
  description: string;
  updated_at: string;
}

export interface FeatureFlagConfig {
  use_livekit_sfu: boolean;
  use_webrtc_mesh: boolean;
  enable_sfu_logs: boolean;
}

export interface MeetingFlagRequest {
  meeting_id: string;
  enabled: boolean;
}

export interface MigrationStats {
  livekit_meeting_count: number;
  global_livekit_enabled: boolean;
  global_mesh_enabled: boolean;
  sfu_logs_enabled: boolean;
  migration_phase: "mesh_only" | "parallel_testing" | "sfu_only" | "unknown";
  updated_at: string;
}

export interface MigrationPhase {
  phase: "mesh_only" | "parallel_testing" | "sfu_only" | "unknown";
  description: string;
  recommendations: string[];
  config: FeatureFlagConfig;
}

export interface SetFlagRequest {
  flag_name: string;
  enabled: boolean;
}

export interface ArchitectureInfo {
  meeting_id: string;
  should_use: boolean;
  architecture: "SFU (LiveKit)" | "Mesh (WebRTC)";
}

class FeatureFlagService {
  private baseUrl = "/api/v1/feature-flags";

  // Get current feature configuration
  async getFeatureConfig(): Promise<FeatureFlagConfig> {
    const response = await apiClient.request<{
      success: boolean;
      data: FeatureFlagConfig;
    }>(`${this.baseUrl}/config`);
    return response.data;
  }

  // Get all feature flags with their status
  async getAllFlags(): Promise<Record<string, FeatureFlag>> {
    const response = await apiClient.request<{
      success: boolean;
      data: Record<string, FeatureFlag>;
    }>(`${this.baseUrl}/all`);
    return response.data;
  }

  // Set a feature flag
  async setFlag(
    flagName: string,
    enabled: boolean
  ): Promise<{ flag_name: string; enabled: boolean }> {
    const response = await apiClient.request<{
      success: boolean;
      data: { flag_name: string; enabled: boolean };
    }>(`${this.baseUrl}/set`, {
      method: "POST",
      body: JSON.stringify({
        flag_name: flagName,
        enabled: enabled,
      } as SetFlagRequest),
    });
    return response.data;
  }

  // Check if a specific flag is enabled
  async isFlagEnabled(
    flagName: string
  ): Promise<{ flag_name: string; enabled: boolean }> {
    const response = await apiClient.request<{
      success: boolean;
      data: { flag_name: string; enabled: boolean };
    }>(`${this.baseUrl}/${flagName}`);
    return response.data;
  }

  // Set multiple flags at once
  async batchSetFlags(requests: SetFlagRequest[]): Promise<{
    success_count: number;
    error_count: number;
    results: { flag_name: string; enabled: boolean }[];
    errors?: { flag_name: string; error: string }[];
  }> {
    const response = await apiClient.request<{
      success: boolean;
      data: {
        success_count: number;
        error_count: number;
        results: { flag_name: string; enabled: boolean }[];
        errors?: { flag_name: string; error: string }[];
      };
    }>(`${this.baseUrl}/batch-set`, {
      method: "POST",
      body: JSON.stringify(requests),
    });
    return response.data;
  }

  // Get migration statistics
  async getMigrationStats(): Promise<MigrationStats> {
    const response = await apiClient.request<{
      success: boolean;
      data: MigrationStats;
    }>(`${this.baseUrl}/migration/stats`);
    return response.data;
  }

  // Get current migration phase and recommendations
  async getMigrationPhase(): Promise<MigrationPhase> {
    const response = await apiClient.request<{
      success: boolean;
      data: MigrationPhase;
    }>(`${this.baseUrl}/migration/phase`);
    return response.data;
  }

  // Enable LiveKit for a specific meeting
  async enableLiveKitForMeeting(
    meetingId: string
  ): Promise<{ meeting_id: string; enabled: boolean }> {
    const response = await apiClient.request<{
      success: boolean;
      data: { meeting_id: string; enabled: boolean };
    }>(`${this.baseUrl}/meetings/livekit/enable`, {
      method: "POST",
      body: JSON.stringify({
        meeting_id: meetingId,
        enabled: true,
      } as MeetingFlagRequest),
    });
    return response.data;
  }

  // Disable LiveKit for a specific meeting
  async disableLiveKitForMeeting(
    meetingId: string
  ): Promise<{ meeting_id: string; enabled: boolean }> {
    const response = await apiClient.request<{
      success: boolean;
      data: { meeting_id: string; enabled: boolean };
    }>(`${this.baseUrl}/meetings/${meetingId}/livekit`, {
      method: "DELETE",
    });
    return response.data;
  }

  // Check if LiveKit should be used for a specific meeting
  async shouldUseLiveKitForMeeting(
    meetingId: string
  ): Promise<ArchitectureInfo> {
    const response = await apiClient.request<{
      success: boolean;
      data: ArchitectureInfo;
    }>(`${this.baseUrl}/meetings/${meetingId}/livekit`);
    return response.data;
  }

  // Cleanup expired flags
  async cleanupExpiredFlags(): Promise<void> {
    await apiClient.request(`${this.baseUrl}/cleanup`, {
      method: "POST",
    });
  }

  // Helper methods for common operations

  // Check if SFU (LiveKit) should be used globally
  async shouldUseSFUGlobally(): Promise<boolean> {
    const config = await this.getFeatureConfig();
    return config.use_livekit_sfu;
  }

  // Check if mesh WebRTC should be used globally
  async shouldUseMeshGlobally(): Promise<boolean> {
    const config = await this.getFeatureConfig();
    return config.use_webrtc_mesh;
  }

  // Enable parallel testing mode (both SFU and mesh)
  async enableParallelTesting(): Promise<void> {
    await this.batchSetFlags([
      { flag_name: "use_livekit_sfu", enabled: true },
      { flag_name: "use_webrtc_mesh", enabled: true },
    ]);
  }

  // Enable SFU-only mode
  async enableSFUOnlyMode(): Promise<void> {
    await this.batchSetFlags([
      { flag_name: "use_livekit_sfu", enabled: true },
      { flag_name: "use_webrtc_mesh", enabled: false },
    ]);
  }

  // Enable mesh-only mode
  async enableMeshOnlyMode(): Promise<void> {
    await this.batchSetFlags([
      { flag_name: "use_livekit_sfu", enabled: false },
      { flag_name: "use_webrtc_mesh", enabled: true },
    ]);
  }

  // Get architecture recommendation for a meeting
  async getArchitectureRecommendation(meetingId: string): Promise<{
    recommended: "SFU (LiveKit)" | "Mesh (WebRTC)";
    reason: string;
    fallback: "SFU (LiveKit)" | "Mesh (WebRTC)";
  }> {
    const architectureInfo = await this.shouldUseLiveKitForMeeting(meetingId);
    const migrationPhase = await this.getMigrationPhase();

    const recommended = architectureInfo.architecture;
    const fallback = architectureInfo.should_use
      ? "Mesh (WebRTC)"
      : "SFU (LiveKit)";

    let reason = "";
    switch (migrationPhase.phase) {
      case "mesh_only":
        reason = "Currently using mesh WebRTC architecture only";
        break;
      case "parallel_testing":
        reason = "Testing both architectures - using configured preference";
        break;
      case "sfu_only":
        reason = "Using SFU architecture for better scalability";
        break;
      default:
        reason = "Architecture determined by meeting-specific settings";
    }

    return {
      recommended,
      reason,
      fallback,
    };
  }

  // Monitor migration progress
  async getMigrationProgress(): Promise<{
    phase: string;
    completion: number; // 0-100
    nextSteps: string[];
    warnings: string[];
  }> {
    const stats = await this.getMigrationStats();
    const phase = await this.getMigrationPhase();

    let completion = 0;
    let nextSteps: string[] = [];
    let warnings: string[] = [];

    switch (stats.migration_phase) {
      case "mesh_only":
        completion = 0;
        nextSteps = phase.recommendations;
        warnings.push(
          "Still using mesh architecture - consider enabling SFU for testing"
        );
        break;
      case "parallel_testing":
        completion = 50;
        nextSteps = [
          "Monitor performance metrics",
          "Collect user feedback",
          "Gradually increase SFU usage",
        ];
        if (stats.livekit_meeting_count === 0) {
          warnings.push(
            "No meetings using LiveKit yet - start enabling for test meetings"
          );
        }
        break;
      case "sfu_only":
        completion = 100;
        nextSteps = [
          "Monitor system performance",
          "Optimize SFU configuration",
          "Plan for scaling",
        ];
        break;
      default:
        completion = 0;
        warnings.push("Unknown migration phase - check configuration");
    }

    return {
      phase: stats.migration_phase,
      completion,
      nextSteps,
      warnings,
    };
  }
}

export const featureFlagService = new FeatureFlagService();
export default featureFlagService;

File: packages/backend/internal/services/feature_flag_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"github.com/filosofine/gomeet-backend/internal/cache"
	"github.com/redis/go-redis/v9"
	"github.com/sirupsen/logrus"
	"gorm.io/gorm"
)

type FeatureFlagService struct {
	redis     *redis.Client
	db        *gorm.DB
	logger    *logrus.Logger
	cacheRepo *cache.CacheRepository
}

type FeatureFlag struct {
	Name        string    `json:"name"`
	Enabled     bool      `json:"enabled"`
	Description string    `json:"description"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type FeatureFlagConfig struct {
	UseLiveKitSFU bool `json:"use_livekit_sfu"`
	UseWebRTCMesh bool `json:"use_webrtc_mesh"`
	EnableSFULogs bool `json:"enable_sfu_logs"`
}

const (
	// Feature flags
	FeatureUseLiveKitSFU = "use_livekit_sfu"
	FeatureUseWebRTCMesh = "use_webrtc_mesh"
	FeatureEnableSFULogs = "enable_sfu_logs"
	
	// Default values
	DefaultUseLiveKitSFU = false // Start with mesh, gradually enable SFU
	DefaultUseWebRTCMesh = true
	DefaultEnableSFULogs = true
)

func NewFeatureFlagService(redisClient *redis.Client, db *gorm.DB, cacheRepo *cache.CacheRepository) *FeatureFlagService {
	logger := logrus.New()
	logger.SetLevel(logrus.InfoLevel)

	service := &FeatureFlagService{
		redis:     redisClient,
		db:        db,
		logger:    logger,
		cacheRepo: cacheRepo,
	}

	// Initialize default feature flags
	service.initializeDefaultFlags()

	return service
}

// initializeDefaultFlags sets up default feature flags
func (s *FeatureFlagService) initializeDefaultFlags() {
	ctx := context.Background()
	
	defaultFlags := map[string]bool{
		FeatureUseLiveKitSFU: DefaultUseLiveKitSFU,
		FeatureUseWebRTCMesh: DefaultUseWebRTCMesh,
		FeatureEnableSFULogs: DefaultEnableSFULogs,
	}

	for flagName, enabled := range defaultFlags {
		// Only set if not already exists
		exists, err := s.redis.Exists(ctx, s.getFlagKey(flagName)).Result()
		if err != nil {
			s.logger.WithError(err).Error("Failed to check feature flag existence")
			continue
		}

		if exists == 0 {
			if err := s.setFlag(ctx, flagName, enabled); err != nil {
				s.logger.WithError(err).WithField("flag", flagName).Error("Failed to set default feature flag")
			} else {
				s.logger.WithFields(logrus.Fields{
					"flag":    flagName,
					"enabled": enabled,
				}).Info("Set default feature flag")
			}
		}
	}
}

// getFlagKey returns the Redis key for a feature flag
func (s *FeatureFlagService) getFlagKey(flagName string) string {
	return fmt.Sprintf("feature_flag:%s", flagName)
}

// setFlag sets a feature flag value
func (s *FeatureFlagService) setFlag(ctx context.Context, flagName string, enabled bool) error {
	// Store in Redis with 24 hour expiry
	key := s.getFlagKey(flagName)
	if err := s.redis.Set(ctx, key, enabled, 24*time.Hour).Err(); err != nil {
		return fmt.Errorf("failed to set feature flag in Redis: %w", err)
	}

	// Log the change
	s.logger.WithFields(logrus.Fields{
		"flag":    flagName,
		"enabled": enabled,
	}).Info("Feature flag updated")

	return nil
}

// getFlagDescription returns a description for a feature flag
func (s *FeatureFlagService) getFlagDescription(flagName string) string {
	descriptions := map[string]string{
		FeatureUseLiveKitSFU: "Use LiveKit SFU for video conferencing instead of mesh WebRTC",
		FeatureUseWebRTCMesh: "Use traditional mesh WebRTC for video conferencing",
		FeatureEnableSFULogs: "Enable detailed logging for SFU operations",
	}

	if desc, exists := descriptions[flagName]; exists {
		return desc
	}
	return "Feature flag"
}

// IsFlagEnabled checks if a feature flag is enabled
func (s *FeatureFlagService) IsFlagEnabled(flagName string) (bool, error) {
	// Try to get from cache first
	enabled, found, err := s.cacheRepo.GetFeatureFlag(flagName)
	if err != nil {
		s.logger.WithError(err).WithField("flag", flagName).Error("Failed to get feature flag from cache")
	}
	
	if found {
		return enabled, nil
	}

	// Cache miss, get from Redis
	ctx := context.Background()
	key := s.getFlagKey(flagName)

	redisEnabled, err := s.redis.Get(ctx, key).Bool()
	if err == nil {
		// Cache the result
		if cacheErr := s.cacheRepo.SetFeatureFlag(flagName, redisEnabled); cacheErr != nil {
			s.logger.WithError(cacheErr).WithField("flag", flagName).Error("Failed to cache feature flag")
		}
		return redisEnabled, nil
	}

	if err != redis.Nil {
		s.logger.WithError(err).WithField("flag", flagName).Error("Failed to get feature flag from Redis")
	}

	// Fallback to default values
	var defaultValue bool
	switch flagName {
	case FeatureUseLiveKitSFU:
		defaultValue = DefaultUseLiveKitSFU
	case FeatureUseWebRTCMesh:
		defaultValue = DefaultUseWebRTCMesh
	case FeatureEnableSFULogs:
		defaultValue = DefaultEnableSFULogs
	default:
		return false, fmt.Errorf("unknown feature flag: %s", flagName)
	}

	// Cache the default value
	if cacheErr := s.cacheRepo.SetFeatureFlag(flagName, defaultValue); cacheErr != nil {
		s.logger.WithError(cacheErr).WithField("flag", flagName).Error("Failed to cache feature flag default value")
	}

	return defaultValue, nil
}

// SetFlag enables or disables a feature flag
func (s *FeatureFlagService) SetFlag(flagName string, enabled bool) error {
	ctx := context.Background()
	
	// Set in Redis
	if err := s.setFlag(ctx, flagName, enabled); err != nil {
		return err
	}
	
	// Update cache
	if err := s.cacheRepo.SetFeatureFlag(flagName, enabled); err != nil {
		s.logger.WithError(err).WithField("flag", flagName).Error("Failed to update feature flag cache")
	}
	
	return nil
}

// GetFeatureConfig returns the current feature configuration
func (s *FeatureFlagService) GetFeatureConfig() (*FeatureFlagConfig, error) {
	config := &FeatureFlagConfig{}

	// Get each feature flag
	useSFU, err := s.IsFlagEnabled(FeatureUseLiveKitSFU)
	if err != nil {
		return nil, fmt.Errorf("failed to get %s flag: %w", FeatureUseLiveKitSFU, err)
	}
	config.UseLiveKitSFU = useSFU

	useMesh, err := s.IsFlagEnabled(FeatureUseWebRTCMesh)
	if err != nil {
		return nil, fmt.Errorf("failed to get %s flag: %w", FeatureUseWebRTCMesh, err)
	}
	config.UseWebRTCMesh = useMesh

	enableLogs, err := s.IsFlagEnabled(FeatureEnableSFULogs)
	if err != nil {
		return nil, fmt.Errorf("failed to get %s flag: %w", FeatureEnableSFULogs, err)
	}
	config.EnableSFULogs = enableLogs

	return config, nil
}

// GetAllFlags returns all feature flags with their current status
func (s *FeatureFlagService) GetAllFlags() (map[string]*FeatureFlag, error) {
	flags := make(map[string]*FeatureFlag)

	flagNames := []string{
		FeatureUseLiveKitSFU,
		FeatureUseWebRTCMesh,
		FeatureEnableSFULogs,
	}

	for _, flagName := range flagNames {
		enabled, err := s.IsFlagEnabled(flagName)
		if err != nil {
			s.logger.WithError(err).WithField("flag", flagName).Error("Failed to get feature flag")
			continue
		}

		flags[flagName] = &FeatureFlag{
			Name:        flagName,
			Enabled:     enabled,
			Description: s.getFlagDescription(flagName),
			UpdatedAt:   time.Now(), // We could store this in Redis for more accuracy
		}
	}

	return flags, nil
}

// EnableLiveKitForMeeting enables LiveKit SFU for a specific meeting
func (s *FeatureFlagService) EnableLiveKitForMeeting(meetingID string) error {
	ctx := context.Background()
	key := fmt.Sprintf("meeting_livekit:%s", meetingID)
	
	if err := s.redis.Set(ctx, key, true, 24*time.Hour).Err(); err != nil {
		return fmt.Errorf("failed to enable LiveKit for meeting: %w", err)
	}

	s.logger.WithField("meeting_id", meetingID).Info("LiveKit enabled for specific meeting")
	return nil
}

// DisableLiveKitForMeeting disables LiveKit SFU for a specific meeting
func (s *FeatureFlagService) DisableLiveKitForMeeting(meetingID string) error {
	ctx := context.Background()
	key := fmt.Sprintf("meeting_livekit:%s", meetingID)
	
	if err := s.redis.Del(ctx, key).Err(); err != nil {
		return fmt.Errorf("failed to disable LiveKit for meeting: %w", err)
	}

	s.logger.WithField("meeting_id", meetingID).Info("LiveKit disabled for specific meeting")
	return nil
}

// ShouldUseLiveKitForMeeting checks if LiveKit should be used for a specific meeting
func (s *FeatureFlagService) ShouldUseLiveKitForMeeting(meetingID string) (bool, error) {
	ctx := context.Background()
	key := fmt.Sprintf("meeting_livekit:%s", meetingID)

	// Check if meeting has specific LiveKit setting
	enabled, err := s.redis.Get(ctx, key).Bool()
	if err == nil {
		return enabled, nil
	}

	if err != redis.Nil {
		s.logger.WithError(err).WithField("meeting_id", meetingID).Error("Failed to check LiveKit setting for meeting")
	}

	// Fallback to global feature flag
	return s.IsFlagEnabled(FeatureUseLiveKitSFU)
}

// GetMigrationStats returns statistics about the migration progress
func (s *FeatureFlagService) GetMigrationStats() (map[string]interface{}, error) {
	ctx := context.Background()
	
	// Count meetings with LiveKit enabled
	pattern := "meeting_livekit:*"
	keys, err := s.redis.Keys(ctx, pattern).Result()
	if err != nil {
		s.logger.WithError(err).Error("Failed to get LiveKit meeting keys")
	}

	liveKitMeetingCount := len(keys)

	// Get global flags
	config, err := s.GetFeatureConfig()
	if err != nil {
		return nil, fmt.Errorf("failed to get feature config: %w", err)
	}

	stats := map[string]interface{}{
		"livekit_meeting_count": liveKitMeetingCount,
		"global_livekit_enabled": config.UseLiveKitSFU,
		"global_mesh_enabled":    config.UseWebRTCMesh,
		"sfu_logs_enabled":       config.EnableSFULogs,
		"migration_phase":       s.getMigrationPhase(config),
		"updated_at":            time.Now(),
	}

	return stats, nil
}

// getMigrationPhase determines the current migration phase based on feature flags
func (s *FeatureFlagService) getMigrationPhase(config *FeatureFlagConfig) string {
	if !config.UseLiveKitSFU && config.UseWebRTCMesh {
		return "mesh_only"
	}
	if config.UseLiveKitSFU && config.UseWebRTCMesh {
		return "parallel_testing"
	}
	if config.UseLiveKitSFU && !config.UseWebRTCMesh {
		return "sfu_only"
	}
	return "unknown"
}

// CleanupExpiredFlags cleans up expired feature flags and meeting-specific settings
func (s *FeatureFlagService) CleanupExpiredFlags() error {
	ctx := context.Background()
	
	// Clean up expired meeting-specific LiveKit settings
	pattern := "meeting_livekit:*"
	keys, err := s.redis.Keys(ctx, pattern).Result()
	if err != nil {
		return fmt.Errorf("failed to get meeting keys for cleanup: %w", err)
	}

	deletedCount := 0
	for _, key := range keys {
		ttl, err := s.redis.TTL(ctx, key).Result()
		if err != nil {
			s.logger.WithError(err).WithField("key", key).Error("Failed to get TTL for key")
			continue
		}

		if ttl == -1 { // No expiry set, clean it up
			if err := s.redis.Del(ctx, key).Err(); err != nil {
				s.logger.WithError(err).WithField("key", key).Error("Failed to delete expired key")
			} else {
				deletedCount++
			}
		}
	}

	s.logger.WithField("deleted_count", deletedCount).Info("Cleaned up expired feature flags")
	return nil
}